examples/stream/src/lib.rs

```diff
@@ -0,0 +1,517 @@
+    ) {
+        match event {
+            ConnectionEvent::FullyNegotiatedInbound(FullyNegotiatedInbound {
+                protocol: (stream, protocol),
+                info: (),
+            }) => {
+                match self.supported_protocols.get_mut(&protocol) {
+                    Some(sender) => match sender.try_send((self.remote, stream)) {
+                        Ok(()) => {}
+                        Err(e) if e.is_full() => {
+                            tracing::debug!(%protocol, "channel is full, dropping inbound stream");
+                        }
+                        Err(e) if e.is_disconnected() => {
+                            // TODO: Remove the sender from the hashmap here?
+                            tracing::debug!(%protocol, "channel is gone, dropping inbound stream");
+                        }
+                        _ => unreachable!(),
+                    },
+                    None => {
+                        // TODO: Can this be hit if the hashmap gets updated whilst we are negotiating?
+                        debug_assert!(
+                            false,
+                            "Negotiated an inbound stream for {protocol} without having a sender"
+                        );
+                    }
+                }
+            }
+            ConnectionEvent::FullyNegotiatedOutbound(FullyNegotiatedOutbound {
+                protocol: (stream, actual_protocol),
+                info: (),
+            }) => {
+                let Some((expected_protocol, sender)) = self.pending_upgrade.take() else {
+                    debug_assert!(
+                        false,
+                        "Negotiated an outbound stream without a back channel"
+                    );
+                    return;
+                };
+                debug_assert_eq!(expected_protocol, actual_protocol);
+
+                let _ = sender.send(Ok(stream));
+            }
+            ConnectionEvent::DialUpgradeError(DialUpgradeError { error, info: () }) => {
+                let Some((_, sender)) = self.pending_upgrade.take() else {
+                    debug_assert!(
+                        false,
+                        "Received a `DialUpgradeError` without a back channel"
+                    );
+                    return;
+                };
+
+                let error = match error {
+                    swarm::StreamUpgradeError::Timeout => {
+                        Error::Io(io::Error::from(io::ErrorKind::TimedOut))
+                    }
+                    swarm::StreamUpgradeError::Apply(v) => void::unreachable(v),
+                    swarm::StreamUpgradeError::NegotiationFailed => Error::UnsupportedProtocol,
+                    swarm::StreamUpgradeError::Io(io) => Error::Io(io),
+                };
+
+                let _ = sender.send(Err(error));
+            }
+            _ => {}
+        }
+    }
+}
+
+pub struct Upgrade {
+    supported_protocols: Vec<StreamProtocol>,
+}
```

Comment:
Maybe worth abstracting into a `libp2p_swarm::upgrade::ReadyMany`?

Reply:
I am a bit torn. I think it could be useful but at the same time, I don't think it is a step in the right direction to expose utilities from `libp2p-swarm`.